### PR TITLE
Migrate JSON data versions as application changes

### PR DESCRIPTION
**Pull request by adammahmood**
### What is the context of this PR?
- Migrates JSON data versions as application changes.
- Allows us to transform a questionnaire over time.
- Migrations are run (if needed) on every request via middleware

### How to review 
- Ensure tests pass
- Ensure the correct migrations are run ie. already executed migrations are not run again
- Ensure you can create migrations as per README instructions

### CODE REVIEW

**Comment by adammahmood** [comment]
[Migrate JSON data versions as application changes. (9)](https://trello.com/c/k9aPKt4X/836-migrate-json-data-versions-as-application-changes-9)

File: eq-author-api/package.json [hunk 1]
```diff
@@ -53,7 +54,8 @@
       "!scripts/*",
       "!middleware/import.js",
       "!app.js",
-      "!utils/datastoreFileSystem.js"
+      "!utils/datastoreFileSystem.js",
+      "!utils/migrations/*"
```
Comment (tgandrews) [inline eq-author-api/package.json]: I would say we should cover this

File: eq-author-api/utils/migrations/createMigration.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+const fs = require("fs").promises;
+
+if (process.argv.length < 3) {
+  /* eslint-disable-next-line no-console */
+  console.info(
+    "\x1b[33m",
```
Comment (tgandrews) [inline eq-author-api/utils/migrations/createMigration.js]: Could we use [chalk[(https://www.npmjs.com/package/chalk) instead

File: eq-author-api/package.json [hunk 3]
```diff
@@ -5,11 +5,12 @@
...
   "scripts": {
     "start": "node app.js",
-    "start:dev": "NODE_ENV=development nodemon --ignore 'data/' --inspect=0.0.0.0:5858",
+    "start:dev": "NODE_ENV=development nodemon --ignore 'data/ migrations/' --inspect=0.0.0.0:5858",
     "lint": "eslint .",
     "test": "./scripts/test.sh",
     "test:breakingChanges": "node scripts/checkForBreakingChanges.js",
-    "dynamodb-admin": "AWS_ACCESS_KEY_ID=dummy AWS_SECRET_ACCESS_KEY=dummy DYNAMO_ENDPOINT=http://localhost:8050 dynamodb-admin"
+    "dynamodb-admin": "AWS_ACCESS_KEY_ID=dummy AWS_SECRET_ACCESS_KEY=dummy DYNAMO_ENDPOINT=http://localhost:8050 dynamodb-admin",
+    "create-migration": "node utils/migrations/createMigration.js"
```
Comment (tgandrews) [inline eq-author-api/package.json]: Maybe this should live in scripts?

File: eq-author-api/README.md [hunk 4]
```diff
@@ -9,6 +9,7 @@ application.
 - [Environment Variables](#environment-variables)
 - [Run using Docker](#run-using-docker)
 - [Tests](#tests)
+- [Migrations](#migrations)
```
Comment (tgandrews) [inline eq-author-api/README.md]: I guess this will impact the docs from @jchapman68 so there may be some merging or a different format to follow

File: eq-author-api/utils/migrations/createMigration.js (hunk 2 above)
Comment (adammahmood) [reply eq-author-api/utils/migrations/createMigration.js]: I considered this but refrained from adding another dependancy. I'm happy to add Chalk if others are in agreement.

Comment (tgandrews) [reply eq-author-api/utils/migrations/createMigration.js]: As it won't be bundled I don't think its a big deal and I personally have no idea what that escape sequence does.

File: eq-author-api/package.json (hunk 1 above)
Comment (adammahmood) [reply eq-author-api/package.json]: Once I move this to `/scripts`, it won't be covered due the existing ignore rule.  We don't test other scripts, why would we test this?

Comment (tgandrews) [reply eq-author-api/package.json]: Yeah, sorry I missed that migrations end up as `/migrations` and not `utils/migrations` so I wanted to make sure those were covered.

File: eq-author-api/scripts/createMigration.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+const fs = require("fs").promises;
+const chalk = require("chalk");
+
+if (process.argv.length < 3) {
+  /* eslint-disable-next-line no-console */
+  console.info(
+    chalk.green(
```
Comment (tgandrews) [inline eq-author-api/scripts/createMigration.js]: Maybe red instead so that its an obvious error?

File: eq-author-api/scripts/createMigration.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+      "Not enough arguments provided to createMigration. Maybe the migration name is missing? Please refer to README."
+    )
+  );
+  return;
```
Comment (tgandrews) [inline eq-author-api/scripts/createMigration.js]: Can we use `process.exit(1)` so the terminal picks it up that the command failed

File: eq-author-api/scripts/createMigration.js (hunk 5 above)
Comment (tgandrews) [reply eq-author-api/scripts/createMigration.js]: Also can we use `console.error`

File: eq-author-api/middleware/runQuestionnaireMigrations.test.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+      },
+    };
+    migrations = [jest.fn()];
+    let result = await runQuestionnaireMigrations(logger)(migrations)(
+      req,
+      res,
+      next
+    );
+    expect(next).toHaveBeenCalled();
+    expect(result).toBeFalsy();
```
Comment (tgandrews) [inline eq-author-api/middleware/runQuestionnaireMigrations.test.js]: Can we assert on no migrations being run instead of result which does not require this functionality for it to be falsy?

File: eq-author-api/middleware/runQuestionnaireMigrations.test.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    let migrationTwo = jest.fn(() => req.questionnaire);
+    let migrationThree = jest.fn(() => req.questionnaire);
+
+    migrations = [migrationOne, migrationTwo, migrationThree];
+
+    await runQuestionnaireMigrations(logger)(migrations)(req, res, next);
+    expect(next).toHaveBeenCalled();
+    expect(migrationOne).not.toHaveBeenCalled();
+    expect(migrationTwo).toHaveBeenCalledWith(req.questionnaire);
+    expect(migrationThree).toHaveBeenCalledWith(req.questionnaire);
```
Comment (tgandrews) [inline eq-author-api/middleware/runQuestionnaireMigrations.test.js]: Could we have this prove it was the output of `migrationTwo`?

File: eq-author-api/utils/datastoreDynamo.js [hunk 9]
```diff
@@ -56,7 +56,7 @@ const getQuestionnaire = id => {
   });
 };
 
-const MAX_UPDATE_TIMES = 2;
+const MAX_UPDATE_TIMES = 5;
```
Comment (tgandrews) [inline eq-author-api/utils/datastoreDynamo.js]: Why is this necessary?
Does removing the equal fix solve the issue?

Comment (adammahmood) [reply eq-author-api/utils/datastoreDynamo.js]: Unsure at the moment.  Having cypress issues and giving a few things a go!

Comment (tgandrews) [reply eq-author-api/utils/datastoreDynamo.js]: If the isEqual is to blame. Maybe move it to a separate PR?

Comment (adammahmood) [reply eq-author-api/utils/datastoreDynamo.js]: There's UI errors as well 😕 
screenshot 2019-03-08 at 15 59 20

Comment (tgandrews) [reply eq-author-api/utils/datastoreDynamo.js]: Weird.

File: eq-author-api/middleware/runQuestionnaireMigrations.test.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    expect(req.questionnaire.title).toEqual("foo");
+    expect(req.questionnaire.description).toEqual("bar");
+    expect(logger.info).toHaveBeenCalledTimes(2);
+    expect(req.questionnaire.version).toEqual(3);
+  });
+});
```
Comment (samiwel) [inline eq-author-api/middleware/runQuestionnaireMigrations.test.js]: I wonder if it's worth adding a test where one of the migrations throws an error. Just to prove that the version does not get incremented and the saveQuestionnaire function should never be called?

File: eq-author-api/migrations/addVersion.js [hunk 11]
```diff
@@ -0,0 +1,6 @@
+//This is an auto-generated file.  Do NOT modify the method signature.
+
+module.exports = questionnaire => {
+  questionnaire.version = 1;
```
Comment (samiwel) [inline eq-author-api/migrations/addVersion.js]: Do we want to be defensive about this? Or are we happy making the assumption that the version number is the source of truth? e.g. should this be `questionnaire.version = questionnaire.version || 1` or should we error/log a warning if it already has a version?

This is more of an open question that a request to change the above BTW.

File: eq-author-api/scripts/createMigration.js [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+
+const name = process.argv[2];
+const filename = `${name}.js`;
+
+fs.readFile("scripts/migration.template.js").then(data => {
+  fs.writeFile(`migrations/${filename}`, data);
+});
+
+/* eslint-disable-next-line no-console */
+console.info(chalk.green(`${name} migration successfully created.`));
```
Comment (samiwel) [inline eq-author-api/scripts/createMigration.js]: Maybe add a hint here to remember to add it to the `migrations/index.js` array?

File: eq-author-api/migrations/addVersion.js (hunk 11 above)
Comment (tgandrews) [reply eq-author-api/migrations/addVersion.js]: I think we have to assume each migration moves from one state to another. The state before (now) has no version so this adds it. I would be worried where the one with a version got it from.

**Comment by adammahmood** [comment]
Removed the following Codacy Markdown check as it conflicts with Prettier:

>Warn when the spacing between a list item’s bullet and its content

File: eq-author-api/migrations/addVersion.js (hunk 11 above)
Comment (tgandrews) [reply eq-author-api/migrations/addVersion.js]: When is it not undefined? Surely this migration would not have been run in that instance?

**Comment by tgandrews** [comment]
> Removed the following Codacy Markdown check as it conflicts with Prettier:
> 
> > Warn when the spacing between a list item’s bullet and its content

prettier isn't running on markdown is it? So how do these conflict?

**Comment by adammahmood** [comment]
> > Removed the following Codacy Markdown check as it conflicts with Prettier:
> > > Warn when the spacing between a list item’s bullet and its content
> 
> prettier isn't running on markdown is it? So how do these conflict?

Prettier started supporting Markdown from version 1.8:
https://prettier.io/blog/2017/11/07/1.8.0.html

As discussed we should update our root `package.json` to run on all supported files (not just `*.js`)

File: eq-author-api/migrations/addVersion.js (hunk 11 above)
Comment (adammahmood) [reply eq-author-api/migrations/addVersion.js]: As discussed, we will leave this untouched and always set to version 1.
